packages/playwright-test/src/runner/projectUtils.ts

```diff
@@ -181,17 +181,24 @@ async function collectFiles(testDir: string, respectGitIgnore: boolean): Promise
       const gitignore = entries.find(e => e.isFile() && e.name === '.gitignore');
       if (gitignore) {
         const content = await readFileAsync(path.join(dir, gitignore.name), 'utf8');
-        const newRules: Rule[] = content.split(/\r?\n/).map(s => {
-          s = s.trim();
-          if (!s)
+        const newRules: Rule[] = content.split(/\r?\n/).map(line => {
+          line = line.trim();
+          if (!line || line[0] === '#')
             return;
-          // Use flipNegate, because we handle negation ourselves.
-          const rule = new minimatch.Minimatch(s, { matchBase: true, dot: true, flipNegate: true }) as any;
-          if (rule.comment)
-            return;
-          rule.dir = dir;
-          return rule;
-        }).filter(rule => !!rule);
+          const negate = line[0] === '!' || line[0] === '^';
+          // Remove the negation from the rule, because we handle negation ourselves.
+          if (negate)
+            line = line.substring(1);
+          return {
+            dir,
+            negate,
+            match: (s: string, partial?: boolean) => {
+              if (partial)
```

Comment:
I don't think we should ever need partial mode for the gitignore, should we use this as an opportunity to reimplement the filtering logic? It seems like it was created by layers with each layer adding into the logic.